sdcm/provision/gce/virtual_machine_provider.py

```diff
@@ -0,0 +1,140 @@
+import logging
+import random
+from typing import Type
+
+from libcloud.compute.base import Node
+from libcloud.compute.types import NodeState
+
+from sdcm.keystore import KeyStore
+from sdcm.provision.gce.disk_struct_provider import DiskStructProvider, DiskStructArgs
+from sdcm.provision.gce.metadata_provider import MetadataProvider, MetadataArgs
+from sdcm.provision.provisioner import InstanceDefinition, PricingModel
+from sdcm.utils.gce_utils import get_gce_service
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class VMInstanceAlreadyExists(Exception):
+    ...
+
+
+class VMInstanceDestructionUnsuccessful(Exception):
+    ...
+
+
+class VMInstanceRebootUnsuccessful(Exception):
+    ...
+
+
+class VMProvisionError(Exception):
+    ...
+
+
+class VirtualMachineProvider:
+    def __init__(self, region: str,
+                 disk_struct_provider: Type[DiskStructProvider],
+                 metadata_provider: Type[MetadataProvider]):
+        self._region = region
+        self._gce_service = get_gce_service(region=region)
+        self._location = self._get_random_location_in_region()
+        self._service_accounts = KeyStore().get_gcp_service_accounts()
+        self._disk_struct_provider = disk_struct_provider
+        self._metadata_provider = metadata_provider
+        self._vm_node_cache: list[Node] = []
+
+    @staticmethod
+    def _sanitize_tags(tags: dict) -> dict:
+        return {k.lower(): v.lower().replace(".", "-") for k, v in tags.items()}
+
+    # pylint: disable=too-many-arguments
+    def get_or_create_instance(self,
+                               disk_struct_args: DiskStructArgs,
+                               metadata_args: MetadataArgs,
+                               instance_definition: InstanceDefinition,
+                               pricing_model: PricingModel,
+                               params: dict) -> Node:
+        # check if the instance with a given name is already in the cache
+        found_instance = self._get_node_from_cache(node_name=instance_definition.name)
+
+        if found_instance:
+            LOGGER.info("Requested the creation of instance with name %s but a "
+                        "instance with that name already exists. Returning the "
+                        "already existing instance.", instance_definition.name)
+            return found_instance
+
+        instance_params = {
+            "name": instance_definition.name,
+            "size": instance_definition.type,
+            "image": instance_definition.image_id,
+            "ex_network": params.get("gce_network"),
```

Comment:
it should be taken from `instance_definition` - `InstanceDefinition` might need some param for it.

Reply:
Or we can create a `NetworkProvider` for this in a separate task, which I would be in favor of. Right now we don't do much we with GCE networks, but we would probably want to setup / teardown networks for test runs in GCE if possbile, not reuse the same one over and over. @fgelcer WDYT?

Reply:
And this param should rather be part of Provider constructor, as it does not depend on instance

Reply:
> Or we can create a `NetworkProvider` for this in a separate task, which I would be in favor of. Right now we don't do much we with GCE networks, but we would probably want to setup / teardown networks for test runs in GCE if possbile, not reuse the same one over and over. @fgelcer WDYT?

creating a `NetworkProvider` should be great, as we have few network nemesis, that nowadays are supported on AWS only, and it would be great to have that ability for all providers

Reply:
In GCE we have one global network, and only it's name is needed.

 It's not related at all to the network nemesis, where we have multiple network devices on each db node,

Reply:
> In GCE we have one global network, and only it's name is needed.
> 
> It's not related at all to the network nemesis, where we have multiple network devices on each db node,

i understand that, but once we will have a `NetworkProvider` we will most likely be able to have multiple network interfaces in GCE as well

Reply:
Let's create a task for it so we discuss it separately (if we need it at all and how to do it).
For now, make it as part of `GCEProvisioner` constructor argument as it does not depend on instance definition at all.